src/main/java/com/google/firebase/FirebaseApp.java

```diff
@@ -427,28 +338,49 @@ FirebaseService getService(String id) {
    * its own instance of this class. This class is not thread safe. The caller (FirebaseApp) must
    * ensure that methods are called serially.
    */
-  static class TokenRefresher {
+  static class TokenRefresher implements CredentialsChangedListener {
 
     private final FirebaseApp firebaseApp;
-    private ScheduledFuture<Task<GetTokenResult>> future;
+    private ScheduledFuture<Void> future;
+    private boolean closed;
 
     TokenRefresher(FirebaseApp app) {
       this.firebaseApp = checkNotNull(app);
     }
 
+    @Override
+    public final synchronized void onChanged(OAuth2Credentials credentials) throws IOException {
+      if (closed) {
+        return;
+      }
+
+      AccessToken accessToken = credentials.getAccessToken();
+      long refreshDelay = accessToken.getExpirationTime().getTime()
+          - System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(5);
+      if (refreshDelay > 0) {
+        scheduleRefresh(refreshDelay);
```

Comment:
Do we need some kind of protection against having multiple outstanding scheduled refreshes?  Since any token change will cause us to schedule a new one, it seems like we could end up with an infinite number in degenerate cases (somebody just calls refresh on the credential a bunch of times or something).

On that note, what motivated this approach?  It seems more complicated / dangerous to trigger off of credential changes rather than the old approach.

Reply:
The `scheduleRefresh()` helper method cancels the last scheduled refresh event before scheduling a new one. I added this bit of info to the javadoc. Moreover:

1. `TokenRefresher` is now thread-safe. It will only handle one token change event at a time. This is further guaranteed by `GoogleCredentials` whose `refresh()` method and the event notification logic are all synchronized.
2. The new approach is not too far off from what we used to have. We used to cache the token, and run the following bit of code whenever we refreshed the credential:

```
if (currentToken.compareAndSet(oldToken, newToken)) {
      listenersCopy = ImmutableList.copyOf(authStateListeners);
      long refreshDelay  = googleOAuthToken.getExpiryTime() - clock.now()
             - TimeUnit.MINUTES.toMillis(5);
      if (refreshDelay > 0) {
             tokenRefresher.scheduleRefresh(refreshDelay);
      } 
     ....
}
```

Now that token caching and refreshing happens within `GoogleCredentials`, we rely on the `onChanged()` event to schedule future refresh events.

Reply:
Cool, thanks for explaining.